jlm/mlir/frontend/MlirToRvsdg.cpp

```diff
@@ -0,0 +1,256 @@
+  convertRegion(mlirOmega.getRegion(0), rvsdgRegion);
+}
+
+jlm::rvsdg::node *
+MlirToRvsdg::convertLambda(mlir::Operation & mlirLambda, rvsdg::region & rvsdgRegion)
+{
+  // Get the name of the function
+  auto functionNameAttribute = mlirLambda.getAttr(::llvm::StringRef("sym_name"));
+  JLM_ASSERT(functionNameAttribute != nullptr);
+  auto * functionName = static_cast<mlir::StringAttr *>(&functionNameAttribute);
+
+  // A lambda node has only the function signature as the result
+  JLM_ASSERT(mlirLambda.getNumResults() == 1);
+  auto result = mlirLambda.getResult(0).getType();
+
+  if (result.getTypeID() != mlir::rvsdg::LambdaRefType::getTypeID())
+  {
+    throw util::error("The result from lambda node is not a LambdaRefType\n");
+  }
+
+  // Create the RVSDG function signature
+  auto * lambdaRefType = static_cast<mlir::rvsdg::LambdaRefType *>(&result);
+  std::vector<std::unique_ptr<rvsdg::type>> argumentTypes;
+  for (auto argumentType : lambdaRefType->getParameterTypes())
+  {
+    argumentTypes.push_back(convertType(argumentType));
+  }
+  std::vector<std::unique_ptr<rvsdg::type>> resultTypes;
+  for (auto returnType : lambdaRefType->getReturnTypes())
+  {
+    resultTypes.push_back(convertType(returnType));
+  }
+  llvm::FunctionType functionType(std::move(argumentTypes), std::move(resultTypes));
+
+  auto rvsdgLambda = llvm::lambda::node::create(
+      &rvsdgRegion,
+      functionType,
+      functionName->getValue().str(),
+      llvm::linkage::external_linkage);
+
+  JLM_ASSERT(mlirLambda.getRegions().size() == 1);
+  auto lambdaRegion = rvsdgLambda->subregion();
+  auto regionResults = convertRegion(mlirLambda.getRegion(0), *lambdaRegion);
+
+  rvsdgLambda->finalize(*regionResults);
+
+  return rvsdgLambda;
+}
+
+std::unique_ptr<rvsdg::type>
+MlirToRvsdg::convertType(mlir::Type & type)
+{
+  if (type.getTypeID() == mlir::IntegerType::getTypeID())
+  {
+    auto * intType = static_cast<mlir::IntegerType *>(&type);
+    return std::make_unique<rvsdg::bittype>(intType->getWidth());
+  }
+  else if (type.getTypeID() == mlir::rvsdg::LoopStateEdgeType::getTypeID())
+  {
+    return std::make_unique<llvm::loopstatetype>();
+  }
+  else if (type.getTypeID() == mlir::rvsdg::MemStateEdgeType::getTypeID())
+  {
+    return std::make_unique<llvm::MemoryStateType>();
+  }
+  else if (type.getTypeID() == mlir::rvsdg::IOStateEdgeType::getTypeID())
+  {
+    return std::make_unique<llvm::iostatetype>();
+  }
+  else
+  {
+    throw util::error("Type conversion not implemented\n");
+  }
+}
+
+size_t
+MlirToRvsdg::getOperandIndex(mlir::Operation * producer, mlir::Value & operand)
+{
+  // TODO
+  // Is there a more elegant way of getting the index of the
+  // operand of an operation?
+  size_t index = 0;
+  if (producer->getNumResults() > 1)
+  {
+    for (mlir::Value tmp : producer->getResults())
+    {
+      if (tmp == operand)
+      {
+        break;
+      }
+      index++;
+    }
+  }
+  return index;
+}
```

Comment:
I think the need for this method only arises, because you keep track of "operations" in convertBlock instead of keeping track of operands/outputs.

Reply:
I've not made an attempt of changing it to track outputs.